node/src/types/appendable_block.rs

```diff
@@ -6,44 +6,61 @@ use std::{
 use datasize::DataSize;
 use num_traits::Zero;
 use thiserror::Error;
+use tracing::error;
 
 use casper_types::{
-    DeployFootprint, DeployHash, Gas, PublicKey, RewardedSignatures, TimeDiff, Timestamp,
-    TransactionConfig,
+    DeployFootprint, Gas, PublicKey, RewardedSignatures, TimeDiff, Timestamp, TransactionConfig,
+    TransactionFootprint, TransactionHash, TransactionV1Category, TransactionV1Footprint,
 };
 
-use super::BlockPayload;
-use crate::types::{DeployHashWithApprovals, TransactionHashWithApprovals};
+use super::{
+    transaction::{DeployHashWithApprovals, TransactionV1HashWithApprovals},
+    BlockPayload,
+};
+use crate::types::TransactionHashWithApprovals;
 
 const NO_LEEWAY: TimeDiff = TimeDiff::from_millis(0);
 
+#[derive(Debug, Error)]
+pub(crate) enum MismatchType {
+    #[error("legacy deploy with V1 footprint")]
+    LegacyDeployWithV1Footprint,
+    #[error("V1 transaction with legacy footprint")]
+    V1TransactionWithLegacyDeployFootprint,
+}
+
 #[derive(Debug, Error)]
 pub(crate) enum AddError {
     #[error("would exceed maximum transfer count per block")]
     TransferCount,
     #[error("would exceed maximum deploy count per block")]
     DeployCount,
+    #[error("would exceed maximum transaction ({0}) count per block")]
+    TransactionCount(TransactionV1Category),
     #[error("would exceed maximum approval count per block")]
     ApprovalCount,
     #[error("would exceed maximum gas per block")]
     GasLimit,
     #[error("would exceed maximum block size")]
     BlockSize,
-    #[error("duplicate deploy")]
+    #[error("duplicate deploy or transaction")]
     Duplicate,
-    #[error("deploy has expired")]
+    #[error("deploy or transaction has expired")]
     Expired,
     #[error("deploy is not valid in this context")]
     InvalidDeploy,
+    #[error("transaction is not valid in this context")]
+    InvalidTransaction,
+    #[error("footprint type mismatch: {0}")]
+    FootprintTypeMismatch(MismatchType),
 }
 
 /// A block that is still being added to. It keeps track of and enforces block limits.
 #[derive(Clone, Eq, PartialEq, DataSize, Debug)]
 pub(crate) struct AppendableBlock {
     transaction_config: TransactionConfig,
-    deploys: Vec<DeployHashWithApprovals>,
-    transfers: Vec<DeployHashWithApprovals>,
-    deploy_and_transfer_set: HashSet<DeployHash>,
+    transactions: Vec<(TransactionHashWithApprovals, TransactionV1Category)>,
+    transaction_and_transfer_set: HashSet<TransactionHash>,
```

Comment:
I think just `transaction_hashes` or something like that would be less confusing?  The "and_transfer" part was really only useful when we were dealing with deploys which could be transfers or not.